enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,281 @@
+- As an operator, I want my cluster infrastructure to be self healing so I want faulty nodes to be remediated automatically for safety scenarios.
+
+#### Story 6
+- As a multi cluster operator, I want to have a universal user experience for managing the Control Plane in a declarative manner across any cloud provider, bare metal and any flavour of the product that have in common the topology assumed in this doc.
+
+#### Story 7
+- As a developer, I want to be able to deploy the smallest possible cluster to save costs, i.e one instance Control Plane.
+
+### Implementation Details/Notes/Constraints [optional]
+
+To satisfy the goals, motivation and stories above this propose a new CRD and controller that manages the Control Plane as a single entity which supports the following features:
+
+#### Declarative horizontal scaling
+##### Scale out
+1 - The controller always reconciles towards expected number of replicas. This must be an odd number.
+2 - Fetch all existing control plane Machine resources by ownerRef. Adopt any other machine having a targeted label.
+3 - Compare with expected replicas number. If expected is higher than current then:
+4 - Check all owned machines have a backed ready node.
+5 - Check all Etcd members for all owned machines are healthy via Cluster Etcd Operator status signalling.
+6 - If (NOT all Etcd members are healthy OR NOT all owned machines have a backed ready node) then controller short circuits here, log, update status and requeue. Else:
+7 - Choose a failure domain.
+8 - Create new Machine object with a templated spec. Go to 1.
+9 - Cluster Etcd Operator watches the new node. It runs a new Etcd pod on it.
+
+##### Scale in
+1 - The controller always reconciles towards expected number of replicas. This must be an odd number.
+2 - Fetch all existing control plane Machine resources by ownerRef. Adopt any other machine having a targeted label.
+3 - Compare with expected replicas number. If expected is lower than current then:
+4 - Check all owned machines have a backed ready node.
+5 - Check all Etcd members for all owned machines are healthy via Cluster Etcd Operator status signalling.
+6 - If (NOT all Etcd members are healthy OR NOT all owned machines have a backed ready node) then controller short circuits here, log, update status and requeue. Else:
+7 - Remove etcd member.
+8 - Delete machine. Go to 1. (Race between the node going away and Cluster Etcd Operator re-adding the member?)
+
+#### Declarative Vertical scaling (scale out + scale in)
+1 - Watch changes to the Control Plane provideSpec
+2 - Fetch all existing control plane Machine resources by ownerRef. Adopt any other machine having a targeted label.
+3 - Fetch all machines with old providerSpec.
+4 - If any machine has old providerSpec then signal controller as "needs upgrade".
+5 - If any machine has "replaced" annotation then trigger scale in workflow (starting in 4) and requeue. Else:
+6 - Pick oldest machine in more populated domain failure.
```

Comment:
I have mixed feelings about the implications of this on bare metal, where there is some correlation between oldest and least problematic. 

Is there a specific reason for preferring FIFO on cloud instead of LIFO?